### ISSUE TITLE
Cannot apply to subproject, and fail on root project

### ISSUE THREAD

**Issue by gotson** [autonomousapps/dependency-analysis-android-gradle-plugin#218]
**Build scan link**

**Plugin version**

0.52.0

**Gradle version**

6.3

**Android Gradle Plugin (AGP) version**

**Describe the bug**

My project has a root project which doesn't do anything, and a sub-project containing my Kotlin code. It is structured like this because there are other sub-projects in other languages (TypeScript).

When i try to apply the plugin in my Kotlin subproject, i get the following error:
```
An exception occurred applying plugin request [id: 'com.autonomousapps.dependency-analysis', version: '0.52.0']
> Failed to apply plugin [id 'com.autonomousapps.dependency-analysis']
   > You must apply the plugin to the root project. Current project is :komga
```

Which seems fair enough to me.

I then try to add the plugin to my root project, and get the following error:
```
A problem occurred configuring project ':komga'.
> Type org.jetbrains.kotlin.gradle.dsl.KotlinProjectExtension not present
```

My root project `build.gradle.kts` is here: https://github.com/gotson/komga/blob/v0.47.0/build.gradle.kts

I modified it like this:
```kotlin
plugins {
  id("com.autonomousapps.dependency-analysis") version "0.52.0"
}

tasks.wrapper {
  gradleVersion = "6.3"
  distributionType = Wrapper.DistributionType.ALL
}
```

My Kotlin subproject `build.gradle.kts` is here: https://github.com/gotson/komga/blob/v0.47.0/komga/build.gradle.kts

**To Reproduce**
1. ...

**Expected behavior**

I would expect the plugin not to crash my Gradle build.

**Additional context**

**Comment by autonomousapps** [comment]
Thank you for this issue and the link to your project. I will take a look.

**Comment by autonomousapps** [comment]
As I suspected from the error you pasted, the Kotlin plugin code is simply not on the classpath, which is very weird given that the `:komga` project is applying the Kotlin plugin. I verified this by modifying your root `build.gradle.kts` as follows
```kotlin
buildscript {
  repositories {
    jcenter()
  }
  dependencies {
    classpath("org.jetbrains.kotlin:kotlin-gradle-plugin:1.3.72")
  }
}
plugins {
  id("com.autonomousapps.dependency-analysis") version "0.52.0"
}
```
and then in `komga/build.gradle.kts`, I commented out the versions for kotlin-jvm and kotlin-kapt (plugins whose code is already on the classpath cannot set a version).

After that, `buildHealth` would run, but then I saw this error relating to jooq: https://scans.gradle.com/s/ucvg2iodpow2s/failure

I'm not sure why that would happen.

(I am investigating if there's a solution to the first problem that I can implement at the level of my plugin.)

**Comment by autonomousapps** [comment]
Ok, I think that the most canonical solution is actually this:
```kotlin
// root build.gradle.kts
plugins {
  kotlin("jvm") version "1.3.72" apply false
  kotlin("plugin.spring") version "1.3.72" apply false
  kotlin("kapt") version "1.3.72" apply false
  id("com.autonomousapps.dependency-analysis") version "0.52.0"
}
```
```kotlin
// komga/build.gradle.kts
plugins {
  kotlin("jvm") 
  kotlin("plugin.spring") 
  kotlin("kapt")
  id("org.springframework.boot") version "2.2.6.RELEASE"
  id("com.github.ben-manes.versions") version "0.28.0"
  id("com.gorylenko.gradle-git-properties") version "2.2.2"
  id("com.rohanprabhu.kotlin-dsl-jooq") version "0.4.5"
  id("org.flywaydb.flyway") version "6.4.0"
  id("com.github.johnrengelman.processes") version "0.5.0"
  id("org.springdoc.openapi-gradle-plugin") version "1.3.0"
  jacoco
}
```
note that the three Kotlin plugins no longer have a version, since it is specified by the root script. I don't have a link to the docs handy, but I believe this is Gradle's preferred pattern.

**Comment by gotson** [comment]
Thanks a lot for taking the time to try some workarounds. I will give it a try later.

Would it make sense to be able to apply your plugin on subprojects only?

**Comment by autonomousapps** [comment]
It's not really about applying the plugin in the root. It's a classloader hierarchy issue (although I'm waiting on word from Gradle on a forum about that to confirm). It seems that my plugin might be in a parent classloader to the one that has the kotlin plugin applied, and so can't see those types. 

That said, I think you're right that we wouldn't see this if the plugin were only applied on subprojects. Unfortunately, that would require a lot of work to re-architect how the extensions work. 
GitHub <notifications@github.com> wrote:
“Thanks a lot for taking the time to try some workarounds. I will give it a try later.

Would it make sense to be able to apply your plugin on subprojects only?”

**Comment by gotson** [comment]
Hi, I tried the latest `0.53.0` but it fails:

```log
> Task :komga:analyzeClassUsageMain FAILED
Caching disabled for task ':komga:analyzeClassUsageMain' because:
  Build cache is disabled
Task ':komga:analyzeClassUsageMain' is not up-to-date because:
  Task has failed previously.
file or directory '/Users/redacted/Development/komga/komga/build/classes/java/test', not found
:komga:analyzeClassUsageMain (Thread[Execution worker for ':' Thread 3,5,main]) completed. Took 0.097 secs.

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':komga:analyzeClassUsageMain'.
> There was a failure while executing work items
   > A failure occurred while executing com.autonomousapps.tasks.ClassListAnalysisWorkAction
      > java.nio.ByteBuffer.limit(I)Ljava/nio/ByteBuffer;
```

My project is 100% kotlin, so it's not surprising there's nothing in `java/test`

**Comment by autonomousapps** [comment]
Thanks for the report. I'll take a look.

**Comment by autonomousapps** [comment]
I tried replicating and ran into an error with jooq. Can you take a look? https://scans.gradle.com/s/mlrv6vesyeo4e

**Comment by gotson** [comment]
Which version of Java are you using ? Can you try with Java 8 ?

**Comment by autonomousapps** [comment]
That was actually the first thing I thought of, but I had the same error with Java 8 https://scans.gradle.com/s/z5d4bqglblsde#infrastructure

**Comment by autonomousapps** [comment]
I believe this is the same issue as https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/issues/224, which was just resolved by publishing 0.53.1. Closing, but feel free to reopen if you encounter it again.

**Comment by gotson** [comment]
Thanks, it's fixed. I have another problem, i will open another issue.
